src/shared/bpf-object.c

```diff
@@ -0,0 +1,139 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/resource.h>
+#include <unistd.h>
+
+#include "bpf-object.h"
+#include "bpf-program-v2.h"
+#include "fd-util.h"
+#include "set.h"
+
+int bpf_object_new(const unsigned char *mem_buf, size_t size, struct bpf_object **object) {
+        _cleanup_(bpf_object_freep) struct bpf_object *p = NULL;
+        /* Standard error code. */
+        long err;
+
+        assert(mem_buf);
+        assert(size > 0);
+        assert(object);
+
+        DECLARE_LIBBPF_OPTS(
+                        bpf_object_open_opts,
+                        opts);
+
+        p = bpf_object__open_mem(mem_buf, size, &opts);
+        /* Should either point to a valid BPF object or store standard error code. */
+        if (!p)
+                return -EINVAL;
+
+        err = libbpf_get_error(p);
+        if (err)
+                return -err;
+
+        *object = TAKE_PTR(p);
+        return 0;
+}
+
+struct bpf_object *bpf_object_free(struct bpf_object *object) {
```

Comment:
Would it be better to call it `bpf_object_close` as it also closes the object?

Reply:
our destructors are usually called _free(). We sometimes use other names (like _done() or _destroy()) if they only do "partial" destruction, i.e. leave some well-defined resources around.

Reply:
Fine for me, anyway I think this is not relevant anymore with the latest changes in this PR.